keps/sig-cluster-lifecycle/kubeadm/20191229-kubeadm-move-out-of-kk.md

```diff
@@ -0,0 +1,635 @@
+kubeadm. The same cherry pick policy will apply as in k/k.
+
+#### Releasing and tagging
+
+##### Challenges
+
+The current plan is for kubeadm to continue to follow the k8s version scheme
+and release schedule. This will continue to make the skew of kubeadm
+and k8s easy to understand for the users and continue to give motivation
+to the kubeadm developers to get work done on time.
+
+This creates some complications:
+- Automation tools have to build kubeadm as part of the release
+- The k/kubeadm tags and branches need to follow those of k/k
+- k/k tags are not SEMVER, which can cause troubles when importing
+the k/kubeadm repository
+
+Currently kubeadm is released as part of the following k8s artifacts:
+- Node tarballs (Linux / Windows)
+- deb/rpm packages
+
+The kubeadm binary is built using either bazel (amd64 only) or using Docker
+(cross-build, all supported architectures).
+
+The binary is built with the same version as the rest of the released binaries
+like the kubelet. The version tooling resides in k/k under the /hack directory
+and it must be copied to k/kubeadm for consistency. The tooling that perform
+the packaging and building are location in k/release and require manual interaction.
+
+##### Building kubeadm from tarballs
+
+For the initial proposal this will not be supported, but if there
+are plans to distribute the kubeadm source code as part of a k8s release,
+the solution would be to write a static KUBE_VERSION file in the tarball.
+
+If GitHub users download the source code directly from the kubeadm repository
+the build process must be able to detect that this is not a Git repository
+and ask the users to provide a KUBE_VERSION file.
+
+##### Bazel vs go build
+
+Bazel was investigated as an option for building kubeadm from the new
+repository but the benefits are unclear. BUILD files present a maintenance
+challenge and also the bazel WORKSPACE and gazelle files need to be kept
+in sync with k8s. On the other hand building using “go build” can take
+around ~20 sec on a fast machine with fast internet.
+
+The majority voted to not use bazel in the k/kubeadm repository.
+
+If caching is required a build container can be used.
+
+##### Syncing tags and branches between k/k and k/kubeadm
+
+A brief discussion of the #prow maintainers revealed the proposal that
+a post-submit job can be used for that instead of prow plugins.
+
+Supposedly such a job will trigger each time a new branch is created
+or if a tag is pushed. This means that the post-submit can trigger any bash
+script or application to get the latest list of branches and tags in k/k
+and apply the same to the kubeadm repository.
+
+Given kubeadm will potentially move within a single cycle this means that
+not all tags and branches have to be synced - see the options in “Versions
+in the support skew”. So potentially there must be a minimum version
+check in the post-submit job.
+
+Scenarios:
+
+###### k/k creates a new release-x.yy branch
```

Comment:
We should give this a name, like "Beta development phase".

Reply:
as in naming the section:
```
###### Beta development phase
```
?

Reply:
Yes.  You have different scenarios listed, but they play out only at specific times during the release cycle and support lifetime, which makes a difference in how we approach solutions requirements.

Reply:
this scenario is just for the even of creation of the branch, not for the whole period of the beta.

Reply:
My thinking was let's talk about those phases where we know there's unique work needing defined and operationalized/automated and especially where k/k's process is not good today (ie: things it would be better to not repeat).  To me it is an open question of what tags are created, by whom, and in what order, and any other side effects that need to happen as we flow through:

 * a branch creation point
 * intermediate phase after branch creation and before code freeze
 * code freeze
 * ".0" release
 * post-".0" patch release

Reply:
I've worked from an opinionated yet open minded place that what we have on k/k for these phases. as implemented in anago, are not the best.  But kubeadm (eg: https://github.com/kubernetes/kubeadm/issues/1963) follows closely what happens on k/k for better or worse.  It's reasonable to not saddle kubeadm with fixing all of these problems today and @neolit123 's expressed openness to iteratively moving things in kubeadm if/when that makes sense, and if/when k/k makes improvements.